server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java

```diff
@@ -164,113 +175,128 @@ public void getSynonymRules(String resourceName, int from, int size, ActionListe
-                        .toArray(SynonymRule[]::new);
-                    listener.onResponse(new PagedResult<>(totalSynonymRules, synonymRules));
-                }
-
-                @Override
-                public void onFailure(Exception e) {
-                    final Throwable cause = ExceptionsHelper.unwrapCause(e);
-                    if (cause instanceof IndexNotFoundException) {
-                        listener.onFailure(new ResourceNotFoundException("Synonym set [" + resourceName + "] not found"));
-                        return;
-                    }
-                    listener.onFailure(e);
+            .execute(new DelegatingIndexNotFoundActionListener<>(resourceName, listener, (l, searchResponse) -> {
+                final long totalSynonymRules = searchResponse.getHits().getTotalHits().value;
+                if (totalSynonymRules == 0) {
+                    l.onFailure(new ResourceNotFoundException("Synonym set [" + resourceName + "] not found"));
+                    return;
                 }
-            });
+                final SynonymRule[] synonymRules = Arrays.stream(searchResponse.getHits().getHits())
+                    .map(SynonymsManagementAPIService::hitToSynonymRule)
+                    .toArray(SynonymRule[]::new);
+                l.onResponse(new PagedResult<>(totalSynonymRules, synonymRules));
+            }));
     }
 
     private static SynonymRule hitToSynonymRule(SearchHit hit) {
-        return new SynonymRule(
-            externalSynonymRuleId(hit.getId()),
-            (String) hit.getSourceAsMap().get(SynonymRule.SYNONYMS_FIELD.getPreferredName())
-        );
-    }
-
-    // Retrieves the external synonym rule ID from the internal one for displaying to users
-    private static String externalSynonymRuleId(String internalId) {
-        int index = internalId.indexOf(SYNONYM_RULE_ID_SEPARATOR);
-        if (index == -1) {
-            throw new IllegalStateException("Synonym Rule ID [" + internalId + "] is incorrect");
-        }
-        return internalId.substring(index + 1);
+        Map<String, Object> sourceAsMap = hit.getSourceAsMap();
+        return new SynonymRule((String) sourceAsMap.get(SYNONYM_RULE_ID_FIELD), (String) sourceAsMap.get(SYNONYMS_FIELD));
     }
 
-    public void putSynonymsSet(String resourceName, SynonymRule[] synonymsSet, ActionListener<UpdateSynonymsSetResult> listener) {
-        deleteSynonymSetRules(resourceName, listener.delegateFailure((deleteByQueryResponseListener, bulkByScrollResponse) -> {
-            boolean created = bulkByScrollResponse.getDeleted() == 0;
-            final List<BulkItemResponse.Failure> bulkFailures = bulkByScrollResponse.getBulkFailures();
-            if (bulkFailures.isEmpty() == false) {
+    public void putSynonymsSet(
+        String resourceName,
+        SynonymRule[] synonymsSet,
+        ActionListener<SynonymsReloadResult<UpdateSynonymsResultStatus>> listener
```

Comment:
I've found this double generics are more difficult to follow, not sure if it is worth introducing extra generic just for 2 classes.

Reply:
I'm not very satisfied with this either. Probably I'll run through other ideas in separate PRs.